### PR TITLE
bpo-31050: IDLE: Factor GenPage class from ConfigDialog

### PR DESCRIPTION
**Pull request by csabella**
https://bugs.python.org/issue31050

### CODE REVIEW

File: Lib/idlelib/configdialog.py [hunk 1]
```diff
@@ -1845,6 +1845,243 @@ def var_changed_space_num(self, *params):
         changes.add_option('main', 'Indent', 'num-spaces', value)
 
 
+class GenPage(Frame):
+
+    def __init__(self, parent):
+        super().__init__(parent)
+        self.parent = parent
```
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py]: Delete unneeded line. See below.

File: Lib/idlelib/configdialog.py [hunk 2]
```diff
@@ -1845,6 +1845,243 @@ def var_changed_space_num(self, *params):
...
+                frame_help: LabelFrame
+                    frame_helplist: Frame
+                        frame_helplist_buttons: Frame
+                            (*)button_helplist_edit
+                            (*)button_helplist_add
+                            (*)button_helplist_remove
+                        (*)helplist: ListBox
+                        scroll_helplist: Scrollbar
+        """
+        parent = self.parent
```
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py]: Delete and replace 'parent' with 'self' in the 4 statements below.

File: Lib/idlelib/configdialog.py [hunk 3]
```diff
@@ -1845,6 +1845,243 @@ def var_changed_space_num(self, *params):
...
+        self.autosave = tracers.add(
+                IntVar(parent), ('main', 'General', 'autosave'))
+        self.win_width = tracers.add(
+                StringVar(parent), ('main', 'EditorWindow', 'width'))
+        self.win_height = tracers.add(
+                StringVar(parent), ('main', 'EditorWindow', 'height'))
+
+        # Create widgets:
+        # body and section frames.
+        frame = self
```
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py]: Delete and carefully replace '(frame,' with '(self,' in the next four statements.  (I believe that these are the only one that need it.)

File: Lib/idlelib/configdialog.py [hunk 4]
```diff
@@ -1845,6 +1845,243 @@ def var_changed_space_num(self, *params):
...
+        # frame_help.
+        frame_helplist_buttons.pack(side=RIGHT, padx=5, pady=5, fill=Y)
+        frame_helplist.pack(side=TOP, padx=5, pady=5, expand=TRUE, fill=BOTH)
+        scroll_helplist.pack(side=RIGHT, anchor=W, fill=Y)
+        self.helplist.pack(side=LEFT, anchor=E, expand=TRUE, fill=BOTH)
+        self.button_helplist_edit.pack(side=TOP, anchor=W, pady=5)
+        self.button_helplist_add.pack(side=TOP, anchor=W)
+        self.button_helplist_remove.pack(side=TOP, anchor=W, pady=5)
+
+        return frame
```
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py]: Delete return line.

File: Lib/idlelib/configdialog.py [hunk 5]
```diff
@@ -1845,6 +1845,243 @@ def var_changed_space_num(self, *params):
...
+                IntVar(parent), ('main', 'General', 'editor-on-startup'))
+        self.autosave = tracers.add(
+                IntVar(parent), ('main', 'General', 'autosave'))
+        self.win_width = tracers.add(
+                StringVar(parent), ('main', 'EditorWindow', 'width'))
+        self.win_height = tracers.add(
+                StringVar(parent), ('main', 'EditorWindow', 'height'))
+
+        # Create widgets:
+        # body and section frames.
```
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py]: Delete 'body and '

File: Lib/idlelib/configdialog.py [hunk 6]
```diff
@@ -1199,253 +1199,253 @@ def activate_config_changes(self):
...
+#        self.update_help_changes()
+#        self.set_add_delete_state()
+#
+#    def update_help_changes(self):
+#        "Clear and rebuild the HelpFiles section in changes"
+#        changes['main']['HelpFiles'] = {}
+#        for num in range(1, len(self.user_helplist) + 1):
+#            changes.add_option(
+#                    'main', 'HelpFiles', str(num),
+#                    ';'.join(self.user_helplist[num-1][:2]))
```
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py]: With tests passing, delete the entire commented out block above.

File: Lib/idlelib/idle_test/test_configdialog.py [hunk 7]
```diff
@@ -237,36 +237,33 @@ def setUp(self):
...
-        # Mask instance methods used by help functions.
-        d.set = d.set_add_delete_state = Func()
-        d.upc = d.update_help_changes = Func()
+        page = cls.page = dialog.genpage
+        dialog.note.select(page)
+        page.set = page.set_add_delete_state = Func()
+        page.upc = page.update_help_changes = Func()
 
     @classmethod
     def tearDownClass(cls):
```
Comment (terryjreedy) [inline Lib/idlelib/idle_test/test_configdialog.py:254]: Add 'page = cls.page' and simplify next 4 lines.

**Review by terryjreedy — CHANGES_REQUESTED** [review]
Add blurb with
bpo-31004: Factor FontPage(Frame) class from ConfigDialog.
The slightly modified tests continue to pass. 
Patch by Cheryl Sabella.

File: Lib/idlelib/configdialog.py [hunk 8]
```diff
@@ -105,13 +105,13 @@ def create_widgets(self):
...
         self.keyspage = self.create_page_keys()
-        self.genpage = self.create_page_general()
+        self.genpage = GenPage(note)
         self.extpage = self.create_page_extensions()
-        note.add(self.fontpage, text='Fonts/Tabs')
-        note.add(self.highpage, text='Highlights')
-        note.add(self.keyspage, text=' Keys ')
-        note.add(self.genpage, text=' General ')
-        note.add(self.extpage, text='Extensions')
+        note.add(self.fontpage, text='Fonts/Tabs', underline=0)
```
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py]: Revert the underline additions.

File: Lib/idlelib/configdialog.py (hunk 6 above)
Comment (csabella) [reply Lib/idlelib/configdialog.py]: Done

File: Lib/idlelib/configdialog.py (hunk 1 above)
Comment (csabella) [reply Lib/idlelib/configdialog.py]: Done

File: Lib/idlelib/configdialog.py (hunk 2 above)
Comment (csabella) [reply Lib/idlelib/configdialog.py]: Done

File: Lib/idlelib/configdialog.py (hunk 5 above)
Comment (csabella) [reply Lib/idlelib/configdialog.py]: Done.

File: Lib/idlelib/configdialog.py (hunk 3 above)
Comment (csabella) [reply Lib/idlelib/configdialog.py]: Done.

File: Lib/idlelib/configdialog.py (hunk 4 above)
Comment (csabella) [reply Lib/idlelib/configdialog.py]: Done.

File: Lib/idlelib/idle_test/test_configdialog.py (hunk 7 above)
Comment (csabella) [reply Lib/idlelib/idle_test/test_configdialog.py:254]: Done.

File: Lib/idlelib/configdialog.py (hunk 8 above)
Comment (csabella) [reply Lib/idlelib/configdialog.py]: Done.

**Review by terryjreedy — APPROVED** [review]
I tested each widget on the page and checked that the corresponding change had been made in my user config-main.  (Automating this would be the subject of a #24845 spinoff.)

**Pull request merged**